overlord/snapstate/check_snap.go

```diff
@@ -99,40 +99,86 @@ func checkVersion(version string) bool {
 	return true
 }
 
-var openSnapFile = backend.OpenSnapFile
+type ErrSnapNeedsMode struct {
+	Mode snap.ConfinementType
+	Snap string
+}
 
-// checkSnap ensures that the snap can be installed.
-func checkSnap(st *state.State, snapFilePath string, si *snap.SideInfo, curInfo *snap.Info, flags Flags) error {
-	// This assumes that the snap was already verified or --dangerous was used.
+func (e *ErrSnapNeedsMode) Error() string {
+	return fmt.Sprintf("snap %q requires %s or confinement override", e.Snap, e.Mode)
+}
 
-	s, _, err := openSnapFile(snapFilePath, si)
-	if err != nil {
-		return err
-	}
+type ErrSnapNeedsClassicSystem string
 
-	if s.NeedsDevMode() && !flags.DevModeAllowed() {
-		return fmt.Errorf("snap %q requires devmode or confinement override", s.Name())
-	}
-	if s.NeedsClassic() {
-		if !release.OnClassic {
-			return fmt.Errorf("snap %q requires classic confinement which is only available on classic systems", s.Name())
+func (e ErrSnapNeedsClassicSystem) Error() string {
+	return fmt.Sprintf("snap %q requires classic confinement which is only available on classic systems", string(e))
+}
+
+func checkSnapInfoFlags(info *snap.Info, snapst *SnapState, flags Flags) error {
+	switch c := info.Confinement; c {
+	case snap.StrictConfinement, "":
+		// strict is always fine
+		return nil
+	case snap.DevModeConfinement:
+		// --devmode needs to be specified every time (==> ignore snapst)
+		if flags.DevModeAllowed() {
+			return nil
+		}
+		return &ErrSnapNeedsMode{
+			Mode: c,
+			Snap: info.Name(),
 		}
-		if !flags.Classic {
-			return fmt.Errorf("snap %q requires consent to use classic confinement", s.Name())
+	case snap.ClassicConfinement:
+		if flags.Classic {
+			return nil
+		}
+
+		if snapst != nil && snapst.Flags.Classic {
+			return nil
+		}
+
+		return &ErrSnapNeedsMode{
+			Mode: c,
+			Snap: info.Name(),
 		}
+	default:
+		return fmt.Errorf("unknown confinement %q", c)
+	}
+}
+
+func checkSnapInfo(info *snap.Info, snapst *SnapState, flags Flags) error {
```

Comment:
also again the name makes you think it does what checkSnap does, but starting from a info, but actually it does less, so again maybe it needs a slightly different name